### ISSUE TITLE
Scroll Jumps and jarring effect with full windowing mode

### ISSUE THREAD

**Issue by singharpit94** [jupyterlab/jupyterlab#16326]
I am using jupyterlab 4.2.0 and observed that the full windowing mode which is enabled by default causes the native scrollbar jumping and causing jarring effects, Attached a gif for more clarity about the issue.
![windowing-bug](https://github.com/jupyterlab/jupyterlab/assets/7591673/2e966c41-93cf-482b-9be9-49b6b3631de5)

Here is the code I used to create tabular tables in all the cells.

```
import datetime as dt
import numpy as np
import pandas as pd
import panel as pn

np.random.seed(7)
pn.extension('tabulator')

df = pd.DataFrame({
    'int': [1, 2, 3],
    'float': [3.14, 6.28, 9.42],
    'str': ['A', 'B', 'C'],
    'bool': [True, False, True],
    'date': [dt.date(2019, 1, 1), dt.date(2020, 1, 1), dt.date(2020, 1, 10)],
    'datetime': [dt.datetime(2019, 1, 1, 10), dt.datetime(2020, 1, 1, 12), dt.datetime(2020, 1, 10, 13)]
}, index=[1, 2, 3])

df_widget = pn.widgets.Tabulator(df, buttons={'Print': "<i class='fa fa-print'></i>"})
df_widget
```

**Comment by krassowski** [comment]
I can reproduce this with the tabulator panel code. I see some jarring with the windowing mode disabled too but it is of different kind. When windowing is enabled I see that the panel tabulator re-renders each time the cell enters the view.

The panel uses `setInterval` trying to detect if the dependencies are available every 25ms for 200 times (5 seconds total). It appears that if a cell was run when out of view, and user scrolls the notebook so that the cell is in the rendering range (but not necessarily visible) within 5 seconds, panel will populate the output and move the scrollbar slightly, but this may not be the major culprit.

The major issue might instead be that the outputs are re-rendered when cell output visibility changes by panel tabulator; this means that it competes with JupyterLab in its ability to optimize out the widgets (outputs) which are not visible to user.  This cannot be reproduced with simple JS rendering in outputs. The re-rendering can be also seen when switching tabs (independent of windowing mode):

![switching-tag-windowing-off](https://github.com/jupyterlab/jupyterlab/assets/5832902/69e956ee-e980-47ff-b3d5-a5b71951689b)

It calls `redrawTable`, which to make things worse trashes the layout:

![image](https://github.com/jupyterlab/jupyterlab/assets/5832902/ae21a553-88e9-4727-b48c-64c161b85324)

When running the cell with tabulator I also see that the output jumps a long way for a split second before the table renders properly; there is some visible layout trashing too:

![outputs-jump](https://github.com/jupyterlab/jupyterlab/assets/5832902/765bde43-71f5-42db-ad13-c3d35fa27089)

If the panel extension re-renders the output this way each time the cell enters the view, it will lead to annoying jumping of the scrollbar indeed. 

It looks like the jump in cell height starts after `adjust_table_size` panel function is called and stops in consequence of `style_redraw` panel function being called:

![image](https://github.com/jupyterlab/jupyterlab/assets/5832902/dadde408-5d12-4fc1-a4a3-5420829eb98a)

**Comment by krassowski** [comment]
On JupyterLab side, we need to make better job to respond to such jumping height of the output. It is quite tricky to get this right.

We should investigate whether using `content-visibility: hidden` for windowing could solve this problem.

**Comment by singharpit94** [comment]
Thanks @krassowski for taking a look here, the actual problem also exists with our inhouse extensions which outputs rich outputs (React based widgets) which does not have the visibility based logic which in general causes issues even without windowing like in tab switching and all as you have mentioned.

I would request we should gather more feedback and test more the native scrollbar with the windowing mode as I think this issue might not be confined to limited cases.

**Comment by krassowski** [comment]
Looking a bit more into why this might be the case for a React component and not only panel custom logic.

For the case of running the cell (in the panel case) there is a logic which is invoked based on the browser events, like CSS stylesheet getting loaded, rather than on explicit visibility check:

![Screenshot from 2024-05-14 11-52-02](https://github.com/jupyterlab/jupyterlab/assets/5832902/0a7fe653-1f5d-40f6-a054-2309de7c86a5)

For the case of scrolling, there is a function which runs every quarter of a second comparing the height of the cell output and triggering the slow update which causes jitter. I think that because scrolling the windowed notebook in changes heights of outputs from 0 to their actual height this is causing the problematic rerender which causes jitter:

```
this._parent_observer = setInterval((()=>{
      const t = this.el.offsetParent;
      this._offset_parent != t && (this._offset_parent = t,
      null != t && (this.compute_viewport(),
      this.invalidate_layout()))
  }
  ), 250));
```

![Screenshot from 2024-05-14 12-06-56](https://github.com/jupyterlab/jupyterlab/assets/5832902/c4b0f0ad-b4de-4d78-8626-7b9f19bfbf98)

Does your React widget has any interval, any resize observer, or any callback to other browser events such as styles being mounted?

**Comment by singharpit94** [comment]
> Does your React widget has any interval, any resize observer, or any callback to other browser events such as styles being mounted?

Yes, we actually use all the above and it just might be the case that we are affected by them.

It will be hard to share an exact reproducer but one way would be to create a widget on top of https://github.com/schrodinger/fixed-data-table-2 and maybe have some browser observers attached to it, it can help create a reproducer

**Comment by burnpanck** [comment]
Could this be the same issue as #15968 ? In that case, there is no tabulator interface involved though...

**Comment by krassowski** [comment]
I don't think so because this issue is specific to windowing mode. The issue in https://github.com/jupyterlab/jupyterlab/issues/15968 seems to be in the scrolling logic shared between all the modes.

**Comment by krassowski** [comment]
Another datapoint here is that size estimation appears to be off. This is causing the scrollbar to be jumping. For example when starting at the top of the notebook:

![image](https://github.com/user-attachments/assets/c1d1e21f-1430-4692-a639-79d9e6e7833f)

And after scrolling to the bottom:

![image](https://github.com/user-attachments/assets/0b42c8aa-3367-4de6-bf5a-82471712ed2a)

https://github.com/jupyterlab/jupyterlab/blob/a036d6b10084d35436682bbb086324a1ea71eaac/packages/ui-components/src/components/windowedlist.ts#L279-L297

**Comment by krassowski** [comment]
This might be because if a cell changes it's height it resets the cache for height and offset of cells below it:

https://github.com/jupyterlab/jupyterlab/blob/a036d6b10084d35436682bbb086324a1ea71eaac/packages/ui-components/src/components/windowedlist.ts#L531-L536

This would explain why it only affects notebooks with cells which change their height.

**Comment by krassowski** [comment]
> Another workaround would be to allow a few frames delay (say up to 0.5 second) in which the output would pinned to the height it had before rendering began. After the timeout the output would resize smoothly to the height demanded by the output container. This would probably result in a pleasant user experience and could be also extended to fix https://github.com/jupyter-book/jupyterlab-myst/issues/243.

Applying it to the entire height indeed fixes the issue. Using debouncer with pinning height on scroll gives a nice user experience.


https://github.com/user-attachments/assets/9c19fe5f-8c14-403e-8896-32358591ff23



However, the initial rendering of cells with outputs (before the output height was computed) is pretty bad:
- if we do not pin them, the problem remains on the first scroll
- if we use the estimated cell size (which is already used to compute estimated scrollbar size)
  - it only works for some textual outputs (which will be improved with https://github.com/jupyterlab/jupyterlab/pull/16706)
  - the size estimation is poorly callibrated
  - for cells with custom output types user sees truncated output and more cells than necessary (from windowing list overscan) for some time and then the smooth resizing from zero (which is a lot of motion) some time after scrolling stopped.


https://github.com/user-attachments/assets/3f12a11e-b69d-4a2e-b5ba-824867fc942d


The [existing logic for recording sizes of the cells on resize](https://github.com/jupyterlab/jupyterlab/blob/4a6bc6c573e0b3a10b1b23edcf0894aff42aa9c0/packages/ui-components/src/components/windowedlist.ts#L1452-L1484) does account for the necessity to scroll back if target cell position changes due to resize of another cell, but it fails badly if the cell increases in height by more than one screen as it can cause an oscillating movement between two sets of cells which get attached/detached interchangeably if these change outputs heights unpredictably.

The resize observer logic also gives some information about cell height as it renders but it does not know whether this is the final height or just an intermediate state (e.g. with outputs not yet rendered). This means that even the "measured" height might be different from what user would expect (if we pin to this intermediate height, the cell will be shorter than user would expect).

I wonder if a combination of pinning height and observing intersection with the viewport would be the best solution:
- always pin height for cells with "settled" height (we can use a debounce marker to define settled) when scroll motion begins
- for cells with no settled height, during scrolling:
   - pin them to their initial/current height when they leave viewport when scrolling down (i.e. if they are part of an overscan so not yet detached) to avoid these from pushing the content down
   - allow cells rendered for the first time in viewport (and below) to resize (until these settle)

**Comment by burnpanck** [comment]
You provided a nice explanation/illustration of the difficulty with this issue [in a comment to #15968, a related, but apparently different issue](https://github.com/jupyterlab/jupyterlab/issues/15968#issuecomment-2653084982). What strikes me with that explanation is the fact that cells would "change their height" when entering the viewport through scrolling in the first place. In my understanding, the UX of scrolling should be a pure "sliding of the viewport" over content that is fundamentally completely unaware and thus independent of the viewport location. Windowing is supposed to be purely an optimisation to increase responsiveness of the page: Cells that are out of the viewport are removed from the DOM to avoid taking up resources (I assume) and re-added when they are about to enter the viewport. Functionally, none of that should be able to change any heights. If the actual implementation does change heights, then that is an issue with the windowing implementation.

I'm no JS expert, but my instinct tells me that the best way solve this would go about this by never letting the height of off-viewport cells determine the position of on-viewport cells within the viewport. Basically, the page geometry would not expand from the "top of the page", but rather from the center of the viewport. That is not how the DOM works natively, but that is the fundamental challenge with any windowing implementation - if it cannot be done correctly, then maybe there shouldn't be any windowing. One sketch could be to absolutely position the cell "in the center of the viewport" (whatever that means - just choose one) on the page, and let the previous cells "stack above" - I'm sure this can be done with CSS in some way. If, upon scrolling back, you notice that a cell above the cell with absolute positioning did in fact change its height since it was last rendered (not due to scrolling, but due to e.g. that cell currently growing some output), then you would need to synchronously change the absolute position of the center cell and the scroll position - something that can be done without jitter if done correctly due to the synchronous nature of JS, but it would require great care. Probably better not to fix the absolute position of the cell "in the center", but rather the cell touching the upper limit of the viewport: then the scroll adjustments only need to happen when transitioning to pin another cell.

Maybe I'm rambling useless stuff here, after all, I don't know much about the internals of windowing. I'm just thinking that the UX of scrolling can be made exactly well defined, at least with respect to a single horizontal line of a single cell (e.g. the top limit of the currently pinned cell in my description). It should be possible to craft an implementation that exactly ("mathematically") follows that UX, and doesn't rely on "debouncing" or other racy behaviours.

**Comment by burnpanck** [comment]
Actually, I realise that the synchronous scroll position adjustment would probable cause the scroll-bar to change position "under your mouse", which would be a source of scroll jitter. So maybe delay that adjustment until the user lets go of the scroll handle? Obviously, that means if there is all of a sudden a lot more content above the current viewport, the user will reach the top of the scrollbar before reaching the top of the content. But that is probably ok / impossible to fix without crippling windowing: If we don't render above viewport content, we don't know its true size until we see it. So the scrollbar might fundamentally be in the wrong position, and it needs to be corrected eventually. When you let go of the handle is a good moment to do so.

**Comment by krassowski** [comment]
> Cells that are out of the viewport are removed from the DOM to avoid taking up resources (I assume) and re-added when they are about to enter the viewport.

Code cells with outputs are not even removed from the DOM. These are just hidden using `display: none`. Issue is some widget libraries have JS code which listens to `display` and re-render when they see a change. Markdown cells do get removed, and there is no problem, unless these use mermaid plots or pictures from remote URLs which render with a delay or jupyterlab-myst is installed which has unstable rendering.

> Functionally, none of that should be able to change any heights

And it does not for 99% of cases, unless interactive outputs have really odd React logic which does not render in a stable way or has a resize observer/mutation observer itself or loads static assets introducing delays this leads to changing height a few times on entering viewport.

Hence, my current thinking is: detect cases where cell output height changed when rendering for the first time and forbid hiding that output ever.

> Basically, the page geometry would not expand from the "top of the page", but rather from the center of the viewport

Yes, I though about it and like the idea; not how things are implemented right no and does not fully solve the issue caused by  cells resizing the scrollbar.

> Actually, I realise that the synchronous scroll position adjustment would probable cause the scroll-bar to change position "under your mouse", which would be a source of scroll jitter

> Maybe I'm rambling useless stuff here, after all, I don't know much about the internals of windowing. 

No, I do appreciate bouncing ideas.

Yup

> So maybe delay that adjustment until the user lets go of the scroll handle?

ALAS I do not see an API to detect this on native browser scrollbar, which is why I noted https://github.com/jupyterlab/jupyterlab/pull/17132 would need to rely on denouncing heuristics (i.e. how much time passed since user scrolled last time)

**Comment by krassowski** [comment]
Coming back to the `tabulator` reproducer. I was able to fix part of the issue by using soft-hiding for cells which resized too frequently (as explored in https://github.com/jupyterlab/jupyterlab/issues/17331#issuecomment-2676441167).

The problem is that `tabulator` has not only a `ResizeObserver` but also an `IntersectionObserver` which kept firing and triggering re-layout. This lead to frames being partially presented or even dropped with multiple tables were entering/leaving the viewport. The key thing is that after eliminating resizing as a factor, the jitter in scrollbar does not result from the windowing logic itself but from the frames being partially presented and the scrollbar not being synchronized with content in these cases.

Forcing `IntersectionObserver` to not fire when notebook is scrolling helped to alleviate this but I also needed to remove other triggers of the layout.

**Comment by krassowski** [comment]
I wonder though if the causality might be not: "incomplete frame" causes "scrollbar jitter", but "blocking code" causes "incomplete frame" and "blocking code" causes "scrollbar position not synced" causes "scrollbar jitter".

If we could ensure that `onScoll` handler gets fired ASAP even if animation frame is delayed, we could possibly avoid jitter even if there is many nodes on the page; however any blocking code (like in callbacks of `ResizeObserver` and `IntersectionObserver`) would prevent syncing... No, after some testing this does not seem to be the case, at least there is a different issue we need to solve first.

Actually, even with https://github.com/jupyterlab/jupyterlab/pull/17339 I can still reproduce some scrollbar jumping. To reproduce need to set CPU throttling, for me x4 is enough. I see that the `top` value jumps. I believe this must be caused by `getSpan` returning a different value when a cell enters viewrport; however the change in the `top` value only causes the viewport to jump, not the scrollbar; the scrollbar is jumping because of a change in estimated total size of the notebook (`_updateTotalSize` method).

**Comment by krassowski** [comment]
> the scrollbar is jumping because of a change in estimated total size of the notebook (_updateTotalSize method).

It is also jumping because `scrollTop` changes as we are attaching/detaching (or toggling visibility) of cells outside of viewport.

Apparently, _most_ browsers (i.e. excluding Safari) implement scroll anchoring, which adjusts the scroll position to keep the element in place, this is messing up the windowing mode logic. This feature is enabled by default, see:
- https://drafts.csswg.org/css-scroll-anchoring/
- https://developer.mozilla.org/en-US/docs/Web/CSS/overflow-anchor/Guide_to_scroll_anchoring
